### ISSUE TITLE
Airplane does not work on Windows due to invalid character in filenames

### ISSUE THREAD

**Issue by matthijskooijman** [cltrudeau/django-airplane#7]
A friend just tried airplane on a Windows machine, resulting in an "OSError: Invalid Argument" while writing a cached file. It turns out that Windows does not allow the `|` in filenames (see https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file). Maybe something like an underscore or a + could be used instead?

**Comment by cltrudeau** [comment]
Hello Matthijs,

I've re-written how the cache file naming happens. I've now put the URL contents in a file and am using a generic filename instead. It passes my local tests, but I don't have a Windows machine to check it with.

If you're able, grab the "winprob" branch and see if it resolves the problem. If so, I'll publish a new version.

BTW, it still knows how to read the old style, just any new items written use the new mechanism.

**Comment by matthijskooijman** [comment]
Awesome, thanks! Also nice that it can still read the old files, well done.

I looked at the commit, and left some remarks as comments on it (I'm a bit of a nitpick, sorry for that :-p). As for testing, @BrendaH is the friend with Windows, she hopes to have some time to test this coming sunday.

**Comment by BrendaH** [comment]
Matthijs under Brenda's account here. Took a bit more work, but seems to work now. See below.

Seems to be not fixed entirely. First off, you used format string literals, which are only supported from Python 3.6 onwards. Unrelated to the use of Windows, but maybe better to use explicit format calls instead? Here's the ones I used to make things work:

    output = ['{name}:{url}'.format(url=url, name=name) for url,name in url_filename_map.items()]

    raise IOError('could not find file *' + filename + '* for url *' + url +'*')

Secondly, it seems that running `path.exists()`  on an invalid path already throws OSErrors:
```
  File "c:\users\brenda\desktop\django-airplane-winprob\airplane\templatetags\airplanetags.py", line 38, in airplane
    return '/static/%s' % cached_filename(url)
  File "c:\users\brenda\desktop\django-airplane-winprob\airplane\utils.py", line 63, in cached_filename
    if path.exists():
  File "c:\program files\python35\lib\pathlib.py", line 1304, in exists
    self.stat()
  File "c:\program files\python35\lib\pathlib.py", line 1126, in stat
    return self._accessor.stat(self)
  File "c:\program files\python35\lib\pathlib.py", line 371, in wrapped
    return strfunc(str(pathobj), *args)
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'C:\\_D_\\Evolution Events\\Artaxerxes (Django inschrijfsysteem)\\.airplane_cache\\https|3A|2F|2Fcdnjs.cloudflare.com|2Fajax|2Flibs|2F
popper.js|2F1.12.9|2Fumd|2Fpopper.min.js'
```

I fixed that now by catching the OSError in `cache_exists()`, but I'm not so sure that is the nicest way to handle this. I could not find any way to test a filename for validity without accessing it with a quick search, though.

```
def cache_exists(url):
    try:
        cached_filename(url)
    except IOError:
        return False
    # This might happen on invalid filenames
    except IOError:
        return False

    return True
```

With that solved, I got:

```
  File "c:\users\brenda\desktop\django-airplane-winprob\airplane\utils.py", line 27, in read_cache_map
    name, url = line.split(':', 1)
ValueError: not enough values to unpack (expected 2, got 1)
```

Turns out there are empty lines in the `cache_map` file. Looking more closely, it seems these come from reading and rewriting the cache_map, when reading lines from a file, the newline is included. I did a quick fix by modifying `read_cache_map()`:

                    url_filename_map[url.strip()] = name

This might not be perfect, I suppose just the trailing newline should be stripped instead?

**Comment by matthijskooijman** [comment]
Wrt to the newlines, this does touch upon a more broad issue: If a url passed to the templatetag for some reason contains a newline, this would break the cache_map file. This will probably not happen for meaningful urls, but it might be good to check (and maybe return unmodified in such a case?) anyway. Alternatively, switching to something like JSON for storing the map could be an alternative which is quite reliable and easy (and handles all escaping etc automatically).

**Comment by cltrudeau** [comment]
Hi Matthijs,

I've pushed a new commit up. I switched it over to storing JSON like you suggested and fixed a couple of other things.

I'm not following you on what causes the IOError in cache_exists() -- my preference is to leave  the IOError because it indicates something has gone horribly awry with the name of the cached file. This might be related to your example of weird extension names. Do you have a test case that you've actually run across, or are you just being careful in the code review? (Which I appreciate!)

The library only works in conjunction with django's static serve, I suspect anything weird like you described will also cause the mimetype from static serve to be wrong -- which means either way you have a problem.

**Comment by matthijskooijman** [comment]
> I'm not following you on what causes the IOError in cache_exists() -- my preference is to leave the IOError because it indicates something has gone horribly awry with the name of the cached file. This might be related to your example of weird extension names. Do you have a test case that you've actually run across, or are you just being careful in the code review? (Which I appreciate!)

Hm, that's weird. The `IOError` is just the one raised there when the file does not exist. The problem was an `OSError` caused by `path.exists` on an invalid filename. I had intended to modify `cache_exists()` to catch both `IOError` and `OSError`, but somehow just caught `OSError` twice. It *did* seem to work, but maybe that was a coincidence.

In any case, the problem is that when a cached file does not exist in the cache map, it will try to see if the old-style mangled name exist, but `path.exists()` raises the `OSError` shown above in this case.


I just remembered another problem I forgot yesterday: when installing on Windows with `pip install git+https://github.com/cltrudeau/django-airplane.git@winprob`, the git clone for that fails because of the invalid filename of https://github.com/cltrudeau/django-airplane/blob/winprob/extras/sample_site/.airplane_cache/https%7C3A%7C2F%7C2Fmaxcdn.bootstrapcdn.com%7C2Fbootstrap%7C2F3.3.5%7C2Fcss%7C2Fbootstrap.min.css. This is probably not an issue in release version installed from pypy, but maybe consider removing that file (unless it serves as a regression test)?

> I've pushed a new commit up. I switched it over to storing JSON like you suggested and fixed a couple of other things.

Looks good, have not been able to test yet, though.

**Comment by cltrudeau** [comment]
I appreciate the feedback you're giving, but in this case I think I'm going to leave it alone. I believe the cases you're talking about are either caused by 1) using the old filename format, which should now go away, 2) weird URLs which static serve is going to have problems with anyway. In the second case, the best I can do is not cache it, which means it is just as easy to remove it from the {airplane} tag. I have updated the documentation to be clear about this limitation.

As for your comment about cloning, yes, you're right, that isn't in the package, but its presence makes it hard for you to help me test, so I've removed the testing cache from the repo. You should be able to pull it now.

**Comment by matthijskooijman** [comment]
> 1) using the old filename format, which should now go away,

Hm, not sure if you've understood what I've been trying to say. The OSError mentioned above is caused by the *compatibility with the old format*, not the *presence of files in the old format* (since on a Windows system, these old-format files cannot actually exist).

So to make it work at all on Windows, you need to either handle this OSError, or remove the compatibility to read files in the old format.

> 2) weird URLs which static serve is going to have problems with anyway. In the second case, the best I can do is not cache it, which means it is just as easy to remove it from the {airplane} tag. I have updated the documentation to be clear about this limitation.

Make sense, documentation looks good to me.

**Comment by cltrudeau** [comment]
Yep, you're right. I didn't understand. I've wrapped the call in a try/except. It was a pain in the butt to test -- couldn't just mock it because Path is used in a bunch of places. I moved the code to a function and then mocked the function. I believe this will catch the problem you were describing. 

Please take a look, and if it can be tested on Windows that'd be great.

Thanks for all the feedback and interest. Always happy when someone else gets use out of my code.

**Comment by matthijskooijman** [comment]
Code looks good and seems to work on Windows :-)

As for patching Path: Note that it is `Path.exists()` that throws the error, not the creation of the Path object itself. Maybe you *could* patch that without adding this (not-so-pretty) _create_path function?

**Comment by cltrudeau** [comment]
Path is used a lot in the code and it really only needs to be patched in the one place, so ugly or not I'm going to leave it be.

Thanks for all your help on this. Windows problem branch has been merged into master and the 1.1 release has been published.
